src/Tizen.Applications.Common/Tizen.Applications/ApplicationInfo.cs

```diff
@@ -201,27 +217,20 @@ public string ApplicationType
         {
             get
             {
-                IDictionary<string, string> metadata = new Dictionary<String, String>();
-
-                Interop.ApplicationManager.AppInfoMetadataCallback cb = (string key, string value, IntPtr userData) =>
-                {
-                    if (key.Length != 0)
-                    {
-                        if (!metadata.ContainsKey(key))
-                            metadata.Add(key, value);
-                    }
-                    return true;
-                };
+                IDictionary<string, string> metadata = new Dictionary<string, string>();
+                GCHandle handle = GCHandle.Alloc(metadata);
```

Comment:
How about hold only cb object on CGHandle?
```c#
            get
            {
                IDictionary<string, string> metadata = new Dictionary<String, String>();

                Interop.ApplicationManager.AppInfoMetadataCallback cb = (string key, string value, IntPtr userData) =>
                {
                    if (key.Length != 0)
                    {
                        if (!metadata.ContainsKey(key))
                            metadata.Add(key, value);
                    }
                    return true;
                };
                var handle = GCHandle.Alloc(cb, GCHandleType.Pinned);

                IntPtr infoHandle = GetInfoHandle();
                if (infoHandle != IntPtr.Zero)
                {
                    err = Interop.ApplicationManager.AppInfoForeachMetadata(infoHandle, cb, IntPtr.Zero);
                    if (err != Interop.ApplicationManager.ErrorCode.None)
                    {
                        Log.Warn(LogTag, "Failed to get application metadata of " + _applicationId + ". err = " + err);
                    }
                }
                handle.Free();

                return metadata;
            }
```

Reply:
I've tested your suggestion with GCHandleType(Pinned).;
ArgumentException is occurred.

01-01 10:41:54.178+0900 I/DOTNET_LAUNCHER(P12042, T12048): System.ArgumentException: Object contains non-primitive or non-blittable data.
01-01 10:41:54.178+0900 I/DOTNET_LAUNCHER(P12042, T12048):    at System.Runtime.InteropServices.GCHandle.InternalAlloc(Object value, GCHandleType type)
01-01 10:41:54.178+0900 I/DOTNET_LAUNCHER(P12042, T12048):    at System.Runtime.InteropServices.GCHandle.Alloc(Object value, GCHandleType type)

Reply:
Please refer this article
https://blogs.msdn.microsoft.com/cbrumme/2003/05/06/asynchronous-operations-pinning/

>Along the same lines, managed Delegates can be marshaled to unmanaged code, where they are exposed as unmanaged function pointers. Calls on those pointers will perform an unmanaged to managed transition; a change in calling convention; entry into the correct AppDomain; and any necessary argument marshaling. Clearly the unmanaged function pointer must refer to a fixed address. It would be a disaster if the GC were relocating that! This leads many applications to create a pinning handle for the delegate. This is completely unnecessary. The unmanaged function pointer actually refers to a native code stub that we dynamically generate to perform the transition & marshaling. This stub exists in fixed memory outside of the GC heap.

일단, delegate는 pinning할 필요가 없다고 합니다.
Crash가 발생한 원인이 다른곳에 있는지 확인이 필요합니다.

Reply:
>     // Check for ShutDown scenario.  This happens only when we have initiated shutdown 
>     // and someone is trying to call in after the CLR is suspended.  In that case, we
>     // must either raise an unmanaged exception or return an HRESULT, depending on the
>     // expectations of our caller.
>     if (!CanRunManagedCode())
> 

this comment might be helpful.

https://github.com/dotnet/coreclr/blob/master/src/vm/dllimportcallback.cpp#L851

Reply:
@hjhun Could you register this app crash issue to https://github.sec.samsung.net/dotnet/home/issues? With the callstack, It seems that the VM can't execute the managed code in special case.

Reply:
https://github.sec.samsung.net/dotnet/home/issues/252

Reply:
crash callstack과 delegate memory reallocated 이슈와는 관계가 없는것으로 보입니다.

Reply:
Let's abandon this patch and trace issue with https://github.sec.samsung.net/dotnet/home/issues/252